Keysfirst

erwindon opened the pull request:
Changed the main page so that it populates based on the salt-keys. Previously that was done based on the actual minion status retrieval. The updated method always provides a quick load of the page. Even when one or more minions are not reachable.
The visual effect is that the page first shows the key status (accepted/denied/unaccepted/rejected). The accepted state will then be replaced by the actual state of the minion.
Also fixed a mixup between the denied and rejected keys.

erwindon commented:
strange, that is why variable "minionsLoaded" was removed.
The gui should be shown as soon as both jobs and keys are available.
The minionstatus will continue on the background.
Until the minionstatus is available, the screen displays the key status of the active minions, which is of course always "accepted".
just tested in firefox and chrome again --> ok

maerteijn commented:
I r(e)tested this a bit by manually triggering the ``goTo('/')`` and now I see indeed what is happening. It is only not happening when you login, only when you press the 'SaltGUI' button when you inspect a recent job. (at least here, with both Chrome and Firefox).

erwindon commented:
In my case the home screen is properly updated both after login and when clicking on SaltGUI logo.
Any browser cache effects?
Any javascript errors on the login sequence?

maerteijn commented:
I think the difference in behavior is explainable by different salt versions. I see the expected behavior using salt 2018.3.x vs 2017.7.x Can you confirm?

As a sidenote, I see the following with salt 2018.3.x: 

<img width="405" alt="jobs" src="https://user-images.githubusercontent.com/121401/40312655-bf582392-5d13-11e8-8823-9309a50948ad.png">

erwindon commented:
I was on 2018.3.x
reverted to 2017.7.x --> everything still working fine with FF+chrome+edge

on the sidenote: these are the internal jobs that are behind the api calls. since I introduced wheel.key.list_all, I've added that one to the list of ignored jobs in _updateJobs.

maerteijn commented:
Ah, maybe I found it:

<img width="957" alt="hint" src="https://user-images.githubusercontent.com/121401/40359876-803fec6a-5dc4-11e8-87b0-3ab2765ad5f8.png">

erwindon commented on saltgui/static/scripts/routes/home.js:
```diff
@@ -178,6 +178,7 @@ class HomeRoute extends Route {
       i = i + 1;
       if(job.Function === "saltutil.find_job") continue;
       if(job.Function === "grains.items") continue;
+      if(job.Function === "wheel.key.list_all") continue;
```
Yes, I think so too!
But it was introduced by Oliver in July 2016.
So it has nothing to do with this PR.
But too much for a separate branch/PR, so I'll include it

erwindon commented:
So your installation is processing MINION information for which no KEY information is available.
I'll investigate the root cause for that.
(and also add some more error handling anyway)

erwindon commented:
The docker image results in "404 Not Found /// The path '/app' was not found." Any hints?

erwindon commented:
Added simple error handling.
Can you please retest?
The console will now log the cases for which no gui element could be found.
2 possibilities that I can think of:
1) the minion list really has an item that is not in the key list (and will never be in that list)
2) the minion list has an item that is not yet in the key list

maerteijn commented:
<img width="1278" alt="not-found" src="https://user-images.githubusercontent.com/121401/40408535-36d07352-5e68-11e8-9c34-fd09868ac43a.png">

For me to reproduce:
- First login, then it works fine.
- Press logout, login again, then it says it can't find these minions

erwindon commented:
OK, this proves it is possibility #2 "the minion list has an item that is not yet in the key list".
On this system, retrieving the Keys from the Master is actually slower than retrieving the Minion data.
The fact that it happens when you login again should be a coincidence, probably caused by the fact that the minions are now even faster.
In my case, there is always a few minions unreachable, due to scheduled shutdown.
Just a moment while I fix this...

erwindon commented:
Reproduced by simulated fast /minions call.
Now fixed so that return values from /keys and /minions can now be processed in any order.

erwindon commented:
the requested changes have been made, anything else I should do for this PR?

maerteijn commented:
@erwindon Thnx! I will test it later today, if it's ok I'll merge it!
